### PR TITLE
refactor(payment_method): [Klarna] store and populate payment_type for klarna_sdk Paylater in response

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
1. store KlarnaPaymentMethod in payment_method_data 
2. Populate Klarna_sdk payment_type in payment_method_data in response

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
1. Create a Klarna Payment Request via Klarna
```
{
    "amount": 1800,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "customer_id": "cus_NwicjmCJbgafjhjfhHr6",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "three_ds",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "9123456789",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "return_url": "https://duck.com",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "payment_method": "pay_later",
    "payment_method_type": "klarna",
    "payment_experience": "invoke_sdk_client",
    "payment_method_data": {
        "pay_later": {
            "klarna_sdk": {
                "token": "65bfb654-ef2c-6b39-942e-ffdc9a105d3cf"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "city": "Groningen",
            "state": "Groningen",
            "zip": "94105",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "+4901761428434",
            "country_code": "+1"
        }
    },
    "shipping": {
        "address": {
            "line1": "444, Stelar Enclave",
            "line2": "California",
            "line3": "Groningen",
            "city": "Groningen",
            "state": "Groningen",
            "zip": "94105",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "446681800",
            "country_code": "+41"
        },
        "email": "guest@example.com"
    },
    "metadata": {},
    "order_details": [
        {
            "product_name": "Apple iphone 15",
            "quantity": 2,
            "amount": 1800,
            "account_name": "transaction_processing"
        }
    ],
    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "128.0.0.1"
    }
}
```
Complete Payment

2.
In db check for `select * from payment_attempt where payment_id = '{payment_id}';`
Check - payment_method_data   | {"pay_later": {"klarna_sdk": {"payment_type": "pay_by_card"}}}
Screenshot 2024-06-12 at 2 01 37 AM

3. Do payment Retrieve for same payment_id
```
curl --location 'http://localhost:8080/payments/{payment_id}?force_sync=true' \
--header 'Accept: application/json' \
--header 'Klarna-Integrator: string' \
--header 'api-key: {api_key}'
```
Response
```
{
    "payment_id": "pay_gd0c5qfyJouK4j1kU63g",
    "merchant_id": "merchant_1718045496",
    "status": "succeeded",
    "amount": 4900,
    "net_amount": 4900,
    "amount_capturable": 0,
    "amount_received": 4900,
    "connector": "klarna",
    "client_secret": "pay_gd0c5qfyJouK4j1kU63g_secret_WliOmb6y7NqvMbHZRCCr",
    "created": "2024-06-11T20:02:01.432Z",
    "currency": "USD",
    "customer_id": "hyperswitch_sdk_demo_id",
    "customer": {
        "id": "hyperswitch_sdk_demo_id",
        "name": null,
        "email": "hyperswitch_sdk_demo_id@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Hello this is description",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "pay_later",
    "payment_method_data": {
        "pay_later": {
            "klarna_sdk": {
                "payment_type": "pay_by_card"
            }
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "New York",
            "country": "US",
            "line1": "509",
            "line2": "Amsterdam Ave",
            "line3": "alsksoe",
            "zip": "10024",
            "state": "NY",
            "first_name": "Test",
            "last_name": "Person-us"
        },
        "phone": {
            "number": "5551234",
            "country_code": "+1"
        },
        "email": "swangi@gmail.com"
    },
    "billing": {
        "address": {
            "city": null,
            "country": "US",
            "line1": null,
            "line2": null,
            "line3": null,
            "zip": null,
            "state": null,
            "first_name": null,
            "last_name": null
        },
        "phone": null,
        "email": null
    },
    "order_details": [
        {
            "brand": null,
            "amount": 4900,
            "category": null,
            "quantity": 1,
            "product_id": null,
            "product_name": "Apple iphone 15",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "requires_shipping": null
        }
    ],
    "email": "hyperswitch_sdk_demo_id@gmail.com",
    "name": null,
    "phone": null,
    "return_url": "http://localhost:9060/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": "invoke_sdk_client",
    "payment_method_type": "klarna",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "3fd0321b-25d6-4567-8657-83c619f08f45",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": {
        "apple_pay": null,
        "airwallex": null,
        "noon": {
            "order_category": "applepay"
        }
    },
    "feature_metadata": null,
    "reference_id": "3fd0321b-25d6-4567-8657-83c619f08f45",
    "payment_link": null,
    "profile_id": "pro_i50h0FOybhWX6cyx1Zd1",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_VnAl4xXtPDW8EFZ94Qob",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-06-11T20:17:01.432Z",
    "fingerprint": null,
    "browser_info": {
        "language": "en-GB",
        "time_zone": -330,
        "ip_address": "::1",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1728,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 1117,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-06-11T20:02:32.030Z",
    "charges": null,
    "frm_metadata": null
}
```
check for payment_method_data it should have value for the payment_type of klarna_sdk payment

Before this PR 
``` 
"payment_method_data": {
        "pay_later": {},
        "billing": null
    },
```
After this PR
```
 "payment_method_data": {
        "pay_later": {
            "klarna_sdk": {
                "payment_type": "pay_by_card"
            }
        },
        "billing": null
    },
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -3900,11 +3919,23 @@ impl From<AdditionalCardInfo> for CardResponse {
     }
 }
 
+impl From<Option<KlarnaSdkPaymentMethod>> for PaylaterResponse {
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:3922]: you can impl on the Type instead of option

**Review by Narayanbhat166 — APPROVED** [review]

**Comment by Narayanbhat166** [comment]
```json
"payment_method_data": {
        "pay_later": {
            "klarna_sdk": {
                "payment_type": "PayByCard"
            }
        },
        "billing": null
    },
```

Can we have the response `payment_type` sent in `snake_case`?

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/connector/klarna/transformers.rs [hunk 2]
```diff
@@ -84,6 +85,41 @@ pub struct KlarnaPaymentsRequest {
 pub struct KlarnaPaymentsResponse {
     order_id: String,
     fraud_status: KlarnaFraudStatus,
+    authorized_payment_method: AuthorizedPaymentMethod,
```
[the remark this answers is not in the document]
Comment (SamraatBansal) [reply crates/router/src/connector/klarna/transformers.rs:88]: Should be an option
```suggestion
    authorized_payment_method: Option<AuthorizedPaymentMethod>,
```

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -1814,7 +1814,9 @@ pub enum AdditionalPaymentData {
     Wallet {
         apple_pay: Option<ApplepayPaymentMethod>,
     },
-    PayLater {},
+    PayLater {
+        klarna_sdk: Option<KlarnaSdkPaymentMethod>,
+    },
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1819]: Can you ensure that this change is compatible for reverts? Create the payment with new version and try to retrieve it from the older version?

Comment (SamraatBansal) [reply crates/api_models/src/payments.rs:1819]: Do we have an update on this @swangi-kumari ?

File: crates/api_models/src/payments.rs [hunk 4]
```diff
@@ -2787,7 +2795,7 @@ pub enum PaymentMethodDataResponse {
     Card(Box<CardResponse>),
     BankTransfer {},
     Wallet {},
-    PayLater {},
+    PayLater(PaylaterResponse),
```
Comment (SamraatBansal) [inline crates/api_models/src/payments.rs:2798]: We should use Box<> for the structs defined inside the enum variants

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs (hunk 3 above)
Comment (swangi-kumari) [reply crates/api_models/src/payments.rs:1819]: Yes @Narayanbhat166 , I tried creating the payment with new version and was able to retrieve it from the older version. It is compatible.
cc- @SamraatBansal

**Review by SamraatBansal — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/api_models/src/payments.rs (hunk 3 above)
Comment (swangi-kumari) [reply crates/api_models/src/payments.rs:1819]: Tested the same, getting expected response.

psync response:
```
"payment_method_data": {
        "pay_later": {
            "klarna_sdk": null
        },
        "billing": null
    }
```

**Pull request merged**